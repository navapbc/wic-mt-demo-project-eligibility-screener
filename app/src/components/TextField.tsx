import { Trans } from 'next-i18next'
import { ChangeEvent, ReactElement } from 'react'

import Required from '@components/Required'

import { I18nKey } from '@src/types'

export type TextFieldProps = {
  handleChange: (
    e: ChangeEvent<HTMLInputElement> | ChangeEvent<HTMLTextAreaElement>
  ) => void
  id: string
  labelKey: I18nKey
  required?: boolean
  type?: 'input' | 'textarea'
  value: string
}

const TextField = (props: TextFieldProps): ReactElement => {
  const { handleChange, id, labelKey, required, type, value } = props
  let textfield: ReactElement
  if (type === 'textarea') {
    textfield = (
      <textarea
        className="usa-textarea"
        id={id}
        name={id}
        onChange={handleChange}
        role="textbox"
        value={value}
      />
    )
  } else {
    textfield = (
      <input
        className="usa-input"
        id={id}
        name={id}
        onChange={handleChange}
        role="textbox"
        value={value}
      />
    )
  }
  return (
    <>
      <label className="usa-label" htmlFor={id}>
        <Trans i18nKey={labelKey} />
        {required && <Required />}
      </label>
      {textfield}
    </>
  )
}

export default TextField
