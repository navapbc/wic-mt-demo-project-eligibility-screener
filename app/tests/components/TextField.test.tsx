import { render, screen } from '@testing-library/react'
import { ChangeEvent } from 'react'

import TextField from '@components/TextField'

const testProps = {
  handleChange: (
    e: ChangeEvent<HTMLInputElement> | ChangeEvent<HTMLTextAreaElement>
  ) => {},
  id: 'input-id',
  labelKey: 'input label',
  required: false,
  value: '',
}

it('should match snapshot when it is a textfield', () => {
  const { container } = render(<TextField {...testProps} type="input" />)
  expect(container).toMatchSnapshot()
})

it('should match snapshot when it is a textarea', () => {
  const { container } = render(<TextField {...testProps} type="textarea" />)
  expect(container).toMatchSnapshot()
})

it('should be a textfield by default', () => {
  render(<TextField {...testProps} />)
  const element = screen.getByRole('textbox')
  expect(element).toBeInTheDocument()
})

it('should match display required marker if required is true', () => {
  render(<TextField {...testProps} required={true} />)
  const required = screen.getByText('*')
  expect(required).toBeInTheDocument()
})

it('should display the value when it is a textfield', () => {
  render(<TextField {...testProps} type="input" value="textfield value" />)
  const textbox = screen.getByRole('textbox')
  expect(textbox).toHaveValue('textfield value')
})
it('should display the value when it is a textarea', () => {
  render(<TextField {...testProps} type="textarea" value="textarea value" />)
  const textbox = screen.getByRole('textbox')
  expect(textbox).toHaveValue('textarea value')
})
