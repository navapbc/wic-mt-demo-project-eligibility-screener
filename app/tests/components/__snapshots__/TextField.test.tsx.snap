// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot when it is a textarea 1`] = `
[
  <label
    className="usa-label"
    htmlFor="input-id"
  >
    input label
  </label>,
  <textarea
    className="usa-textarea"
    id="input-id"
    name="input-id"
    onChange={[Function]}
    role="textbox"
    value=""
  />,
]
`;

exports[`should match snapshot when it is a textfield 1`] = `
[
  <label
    className="usa-label"
    htmlFor="input-id"
  >
    input label
  </label>,
  <input
    className="usa-input"
    id="input-id"
    name="input-id"
    onChange={[Function]}
    role="textbox"
    value=""
  />,
]
`;
